python/ray/data/_internal/util.py

```diff
@@ -610,3 +622,163 @@ def get_attribute_from_class_name(class_name: str) -> Any:
     module_name = ".".join(paths[:-1])
     attribute_name = paths[-1]
     return getattr(import_module(module_name), attribute_name)
+
+
+class Queue:
```

Comment:
This PR only moves the code from another file. @c21 it seems that you added this initially. any reason why not use Python's built-in queue?

Reply:
@stephanie-wang - this Queue impl is added because Python Queue impls not allow cooperatively exit multiple threads, this PR added this - https://github.com/ray-project/ray/pull/34819 .

For the case of

```
for data in multi_threads_iter():
  if ... :
    break # <-- we should release all holding resource (e.g. last one read data batch) from all threads.
```

Reply:
I added the reason in PR description of https://github.com/ray-project/ray/pull/34819, copied here:

> Change from using Python queue.Queue, to use a set of Semaphore, Lock and plain deque to allow cooperatively exit producer threads. I don't find how to achieve the same by using any classes of Python thread-safe Queues, so roll my own version of producer-consumer queue here.

I can update to add more comments into class later.

Reply:
Ah I see, thanks.